### PR TITLE
Fatma qunnies python 2 task

### PR DESCRIPTION
**Pull request by fatmaQunnies**
The code implements a Task Manager Console Application in Python, utilizing object-oriented principles to manage tasks effectively. It includes features for adding, viewing, updating, and deleting tasks, with robust input validation and error handling to ensure stability and user-friendly interaction.

### CODE REVIEW

File: FatmaQunnies/Task.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+from datetime import date
+
+class Task:
+ 
+    counter = 0  
+    def __init__(self, title, description, due_date, status = 'InProgress'):
```
Comment (Mohammed-samaheen) [inline FatmaQunnies/Task.py:6]: I like the idea of using a default value for the status. However, I wanted to mention something I didn't bring up before: you can use [enums](https://docs.python.org/3/library/enum.html) in Python. This way, if we ever want to update the InProgress value to something like in-progress, we only need to update the enum. But anyway, great work!

File: README.md
Comment (Mohammed-samaheen) [inline README.md:1]: Don't update this README file, instead create a new one in you folder

File: FatmaQunnies/Utils.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import re
+
+def validate_date(date_str):
+    while True:
```
Comment (Mohammed-samaheen) [inline FatmaQunnies/Utils.py:4]: The validation function should focus only on the validation, so user input shouldn't be part of the validation process. Instead, we should store all the user input data as-is (without any validation) in a dictionary with key-value pairs. Then, you can either iterate over that dictionary to run the validation functions OR create a class—let's say UserTaskInputData and perform all the validations within that class.

**Review by Mohammed-samaheen — CHANGES_REQUESTED** [review]

**Comment by Mohammed-samaheen** [comment]
I added [.gitignore](https://github.com/AHL-Logics-backend-tranining-2024/python-task-2/blob/main/.gitignore)

File: FatmaQunnies/Task.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        return displayString
+    
+    
+# task =Task("pythonTask","classes","2021-02-02")
+# task2 =Task("pythonTask2","classes","2021-02-02")
+# task3 =Task("pythonTask3","classes","2021-02-02")
+
+# print(task.display())
+# print(task2.display())
+# print(task3.display())
```
Comment (AMROAMRO404) [inline FatmaQunnies/Task.py:40]: can you please delete these lines?

File: FatmaQunnies/Task_manager.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  is_urgent_input = get_valid_input(is_urgent_inputt).strip().lower()
+  if is_urgent_input in ['yes', 'y','ye']:
+        priority_input = input("Enter the priority (High/Medium/Low)  ").strip()
+        priority = validate_priority(priority_input)
+        task = UrgentTask(title, description, due_date, priority)
+        tasks.append(task)
+  else:
+    task = Task(title, description, due_date)
+    tasks.append(task)
+  print("Task added successfully.")
```
Comment (AMROAMRO404) [inline FatmaQunnies/Task_manager.py:26]: what if this code raise an error, how you would handle it ?

File: FatmaQunnies/Task_manager.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+                title if title else None,
+                description if description else None,
+                due_date if due_date else None,
+                status if status else None
+            )
+            print("Task updated successfully.")
+            break
+
+    
+    print("Task not found.")
```
Comment (AMROAMRO404) [inline FatmaQunnies/Task_manager.py:77]: same thing here, what if this code raise an error?

File: FatmaQunnies/Task_manager.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+            break
+        else:
+            print("Invalid choice. Please try again.")
+Menu()
+# view_tasks()
+# add_task()
+# view_tasks()
+# update_task()
+# view_tasks()
+# delete_task()
```
Comment (AMROAMRO404) [inline FatmaQunnies/Task_manager.py:110]: can you please delete these lines ?

File: FatmaQunnies/Utils.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        else:
+            print("Input cannot be empty. Please try again.")
+ 
+
+def validate_status(status):
+ valid_statuses = ['inProgress', 'completed']
+ while True:
+        # status = input("Enter status (InProgress or Completed): ").strip().lower()
+        # if status in valid_statuses:
+        #     return status
```
Comment (AMROAMRO404) [inline FatmaQunnies/Utils.py:26]: here too

File: FatmaQunnies/Utils.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    while True:
+        # date_str = input(date_str)
+        if re.match(r'^\d{4}-\d{2}-\d{2}$', date_str):
+            return date_str
+        else:
+            print("Invalid date format. Use YYYY-MM-DD. Please try again.")
+
+def get_valid_input(inputt):
+  while True:
+        #user_input = input(inputt).strip()
```
Comment (AMROAMRO404) [inline FatmaQunnies/Utils.py:14]: here too.

File: FatmaQunnies/urgent_task.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+    
+
+
+    def display(self):
+        displayString = (super().display()+f"priority: {self.priority}\n"
+           )
+        return displayString
+    
+urgent_task = UrgentTask("Important Task", "Review ", "2024-08-05", "Medium")
+# print(urgent_task.display())
```
Comment (AMROAMRO404) [inline FatmaQunnies/urgent_task.py:16]: here

File: FatmaQunnies/urgent_task.py [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+        self.priority=priority
+    
+
+
+    def display(self):
+        displayString = (super().display()+f"priority: {self.priority}\n"
+           )
+        return displayString
+    
+urgent_task = UrgentTask("Important Task", "Review ", "2024-08-05", "Medium")
```
Comment (AMROAMRO404) [inline FatmaQunnies/urgent_task.py:15]: here too.

File: FatmaQunnies/Task_manager.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+  description = get_valid_input(description_input)
+  date_str = input("Enter the  date ")
+  due_date =validate_date(date_str)
+  is_urgent_inputt = input("Is this task urgent? (yes/no): ").strip()
+  is_urgent_input = get_valid_input(is_urgent_inputt).strip().lower()
+  if is_urgent_input in ['yes', 'y','ye']:
+        priority_input = input("Enter the priority (High/Medium/Low)  ").strip()
+        priority = validate_priority(priority_input)
+        task = UrgentTask(title, description, due_date, priority)
+        tasks.append(task)
```
Comment (AMROAMRO404) [inline FatmaQunnies/Task_manager.py:22]: can you please update the indentation here ?

File: FatmaQunnies/Task_manager.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+  else:
+    task = Task(title, description, due_date)
+    tasks.append(task)
```
Comment (AMROAMRO404) [inline FatmaQunnies/Task_manager.py:25]: `tasks.append(task)` called twice in this if-else block, maybe we can have only one ?

**Comment by AMROAMRO404** [comment]
maybe you can use prettier for formatting ?

**Comment by AMROAMRO404** [comment]
can you please resolve the confilct ?

File: FatmaQunnies/utils.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
+import re
+from task_status import Status
+
+def validate_date(date_str):
+    while True:
+        if re.match(r'^\d{4}-\d{2}-\d{2}$', date_str):
```
Comment (Mohammed-samaheen) [inline FatmaQunnies/utils.py:6]: you can use `datetime` instead of regex.

File: FatmaQunnies/utils.py [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+            return inputt
+        else:
+            print("Input cannot be empty. Please try again.")
+ 
+
+def validate_status(status):
+ valid_statuses = ['inProgress', 'completed']
+ while True:
+       
+        if status.upper() in Status.__members__:
```
Comment (Mohammed-samaheen) [inline FatmaQunnies/utils.py:24]: Using the upper() method will convert all the letters to uppercase. `inProgress` != `INPROGRESS`

File: FatmaQunnies/utils.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+            return Status[status.upper()]
+        
+        else:
+            print(f"Invalid status. Choose from {valid_statuses}. Please try again.")
+
+def validate_priority(priority):
+    valid_priorities = ['high', 'medium', 'low']
+    while True:
```
Comment (Mohammed-samaheen) [inline FatmaQunnies/utils.py:32]: Am I missing something? Why is there a while loop?

File: FatmaQunnies/utils.py [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+        if priority in valid_priorities:
+            return priority 
+        print(f"Invalid priority. Choose from {valid_priorities}.")
+        priority = input("Enter the priority (high/medium/low) ").strip().lower()
```
Comment (Mohammed-samaheen) [inline FatmaQunnies/utils.py:36]: As mentioned, the input handling should not be included in the validation function.

File: FatmaQunnies/task_manager.py [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+        print("2. View Tasks")
+        print("3. Update Task")
+        print("4. Delete Task")
+        print("5. Exit")
+        
+        try:
+            task_input = input("Enter your choice: ").strip()
+            choice = get_valid_input(task_input)
+            
+            if choice == '1':
```
Comment (Mohammed-samaheen) [inline FatmaQunnies/task_manager.py:94]: you can use match statement.

**Review by Mohammed-samaheen — CHANGES_REQUESTED** [review]

**Review by Mohammed-samaheen — APPROVED** [review]
LGTM

**Review by AMROAMRO404 — APPROVED** [review]

**Pull request merged**